### ISSUE TITLE
[BUG] Azure Table Returned Bad Request When Concurrent Upsert Requests Are Sent

### ISSUE THREAD

**Issue by KokJun-Lye** [Azure/azure-sdk-for-net#37870]
### Library name and version

Azure.Data.Tables 12.8.0

### Describe the bug

When `HttpClient.Timeout` is set in `TableClientOptions` and concurrent upsert requests are sent and with most of the requests are taking the duration that is close to the specified timeout, some of the upserts requests are getting 400 (Bad Request) response from Azure Table.

From the error call stack, the `message.Request.Content` ->`JsonWriter` -> `BytesCommitted` is 0 at this [line](https://github.com/Azure/azure-sdk-for-net/blob/6fb0cd61c366341f2556f97356c0da1fa86766f4/sdk/tables/Azure.Data.Tables/src/Generated/TableRestClient.cs#L765) when the error happened. Also, from the network trace, we can also see that the request content is empty.

Will this be a case when retry and timeout happen concurrently, the `message` object is being disposed prematurely?

### Expected behavior

All upsert requests should be getting 204 (No content) responses from Azure Table.

### Actual behavior

Some of the upserts requests are getting 400 (Bad Request) response from Azure Table.

### Reproduction Steps

Below are not the actual codes but an extraction of them on how the `TableClientOptions` is configured
```
var tableClientOptions = new TableClientOptions();
tableClientOptions.Retry.Mode = RetryMode.Fixed;
tableClientOptions.Retry.Delay = TimeSpan.FromMilliseconds(150);
tableClientOptions.Retry.MaxRetries = 3;
tableClientOptions.Transport = new HttpClientTransport(new HttpClient() { Timeout = TimeSpan.FromMilliseconds(2500) });
 
var tableClient = new TableClient(connectionString, tableName, tableClientOptions);
await tableClient.UpsertEntityAsync(entity, TableUpdateMode.Replace, cancellationToken: cancellationToken).ConfigureAwait(false);
```

We have set a deliberate delay using Fiddler that is close to 2500ms (e.g. 2200ms) and used JMeter to initiate concurrent requests (30 requests per second) to perform the upserts. The 400 responses are returned from Azure Table intermittently.

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by christothes** [comment]
Hi @KokJun-Lye 
Does the behavior reproduce if you set the `tableClientOptions.Retry.NetworkTimeout` property instead of setting a custom transport to specify the timeout?

**Comment by KokJun-Lye** [comment]
> Hi @KokJun-Lye Does the behavior reproduce if you set the `tableClientOptions.Retry.NetworkTimeout` property instead of setting a custom transport to specify the timeout?

Hi @christothes 
Yes, we have also tried setting the `TableClientOptions.Retry.NetworkTimeout` and removed the custom transport:

```
var tableClientOptions = new TableClientOptions();
tableClientOptions.Retry.Mode = RetryMode.Fixed;
tableClientOptions.Retry.Delay = TimeSpan.FromMilliseconds(150);
tableClientOptions.Retry.MaxRetries = 3;
tableClientOptions.Retry.NetworkTimeout =TimeSpan.FromMilliseconds(2500);
```
But the behavior could still be reproduced.

**Comment by christothes** [comment]
Thanks - could you reproduce this without fiddler but with logging enabled and share the output? 
Details on how to enable logging can be found [here.](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/samples/Diagnostics.md)

**Comment by KokJun-Lye** [comment]
> Thanks - could you reproduce this without fiddler but with logging enabled and share the output? Details on how to enable logging can be found [here.](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/samples/Diagnostics.md)

Hi @christothes , it's replicable without Fiddler. If we are testing this in Azure Functions, I guess all we need is configuring the log level for `Azure` as `Trace` right?

**Comment by rafflesgun** [comment]
Hi @christothes,

I just did isolated test using Console App and add Project Reference to `Azure.Data.Tables` &  `Azure.Core` which allow me to add detail log with Console logging.

Here the console output when the issue happen.

![image](https://github.com/Azure/azure-sdk-for-net/assets/5162808/ed6759ff-fb7a-452f-90bd-b9cc7f4a3a6a)

Randomly when `NetworkTimeout` triggered with `OperationCanceledException` in this [line](https://github.com/Azure/azure-sdk-for-net/blob/ec8530d3bad6dc32c4739b7d3f8540cf552d45b4/sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs#L61C17-L61C96), the Request Content become **empty** and subsequence retries would having empty request content which leads to `400 (Bad Request) - InvalidInput` response from Azure Table Api.

400 Response Content from Fiddler
```
HTTP/1.1 400 Bad Request
Transfer-Encoding: chunked
Content-Type: application/json;odata=minimalmetadata;streaming=true;charset=utf-8
Server: Windows-Azure-Table/1.0 Microsoft-HTTPAPI/2.0
x-ms-request-id: 42eb763d-c002-0052-5e46-d0268b000000
x-ms-client-request-id: e8f5b6cc-17b9-40d8-bd9f-55a64ed88287
x-ms-version: 2020-12-06
X-Content-Type-Options: nosniff
Date: Wed, 16 Aug 2023 13:33:22 GMT

{"odata.error":{"code":"InvalidInput","message":{"lang":"en-US","value":"An error occurred while processing this request.\nRequestId:42eb763d-c002-0052-5e46-d0268b000000\nTime:2023-08-16T13:33:23.1393521Z"}}}
```

Code Snippet for Console App
```
 public static async Task Main(string[] args)
        {
            Console.WriteLine("Start Table Test..");

            // Setup Fiddler Proxy for AutoResponder enabled with Rule "METHOD:PUT {resource-name}.table.core.windows.net" and "*delay:2500"
            IWebProxy proxy = new WebProxy("127.0.0.1", 8888);
            proxy.Credentials = CredentialCache.DefaultCredentials;
            HttpClient.DefaultProxy = proxy;

            var tableClientOptions = new TableClientOptions();
            tableClientOptions.Retry.Mode = RetryMode.Fixed;
            tableClientOptions.Retry.Delay = TimeSpan.FromMilliseconds(150);
            tableClientOptions.Retry.MaxRetries = 3;
            tableClientOptions.Retry.NetworkTimeout = TimeSpan.FromMilliseconds(2500);

            var tableClient = new TableClient(connectionString, tableName, tableClientOptions);

            for (var loop = 1; loop <= 2; loop++)
            {
                Console.WriteLine($"Loop: {loop}");

                var taskList = new List<Task>();

                for (var i = 0; i < 25 * loop; i++)
                {
                    taskList.Add(Upsert(tableClient, loop, i));
                }

                try
                {
                    Task.WaitAll(taskList.ToArray());
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Exception: {ex.Message}");
                }

                Console.WriteLine($"End Loop: {loop}");
                Console.WriteLine();
            }

            Console.WriteLine("All Completed");
        }

        public static async Task Upsert(TableClient tableClient, int loop, int index)
        {
            await tableClient.UpsertEntityAsync(
                           new TestTable()
                           {
                               PartitionKey = $"tenant-{loop}-{index}",
                               RowKey = "row-01",
                               StartedTime = DateTimeOffset.UtcNow.AddMinutes(-15),
                               ItemId = Guid.NewGuid().ToString(),
                            },
                           TableUpdateMode.Replace);

            Console.WriteLine($"Responsed: {index}");
        }
```

**Comment by christothes** [comment]
Hi @KokJun-Lye 
This code doesn't fail in my local testing. But seeing a content length of 0 for a response body is normal. It is the request that should have content, but our logging doesn't report request content size. Since you are using Fiddler to capture the requests, can you see the contents of the request body for the one that fails?

**Comment by KokJun-Lye** [comment]
> Hi @rafflesgun This code doesn't fail in my local testing. But seeing a content length of 0 for a response body is normal. It is the request that should have content, but our logging doesn't report request content size. Since you are using Fiddler to capture the requests, can you see the contents of the request body for the one that fails?

Hi @christothes , @rafflesgun will provide an update soon on this.

From my testing, my Fiddler logs showed that the content of the request body is cleared when this happened.

**Comment by rafflesgun** [comment]
> Hi @rafflesgun This code doesn't fail in my local testing. But seeing a content length of 0 for a response body is normal. It is the request that should have content, but our logging doesn't report request content size. Since you are using Fiddler to capture the requests, can you see the contents of the request body for the one that fails?

Hi @christothes,

When the log show content length of 0, the fiddler also having header where 'Content-Length: 0' with empty body for the request.

![image](https://github.com/Azure/azure-sdk-for-net/assets/5162808/4fb68457-d2f1-4f89-9a42-d3c062dc46b3)

**Comment by christothes** [comment]
Unfortunately, I am unable to reproduce this still, even if I increase the number of concurrent operations and lower the timeout. Could you try reproducing this with the latest 12.8.1 version of Azure.Data.Tables?

**Comment by rafflesgun** [comment]
Hi @christothes 
Just did another run this morning with 12.8.1 and this issue still occurred with more occurrence in loop no. 4 (Total loops is 4).
![image](https://github.com/Azure/azure-sdk-for-net/assets/5162808/35a61990-6f3d-494b-8ba9-93dcebdde843)

**Comment by christothes** [comment]
I'm afraid I still can't reproduce this nor can I see how we could ever end up in this state. Is there any factor about your repro that, if changed, prevents you from being able to reproduce?

**Comment by rafflesgun** [comment]
Nothing changed in the Console App beside updated the "Azure.Data.Tables" to version 12.8.1.

The only different from the above code snippet was the Entity i used for testing as follow with more complex property.
``` 
await service.UpsertEntityAsync(
                           new TestTable()
                           {
                               PartitionKey = $"tenant-{loop}-{index}",
                               RowKey = "row-01",
                               Timestamp = DateTimeOffset.UtcNow,
                               StartedTime = DateTimeOffset.UtcNow.AddMinutes(-15),
                               EndedTime = DateTimeOffset.UtcNow.AddMinutes(-5),
                               ItemId_1 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_2 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_3 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_4 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_5 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_6 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_7 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_8 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_9 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               ItemId_10 = $"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}",
                               CustomProperties = new JObject()
                               {
                                   { "some-random-id-0", new JValue(Guid.NewGuid().ToString()) },
                                   { "some-random-id-1", new JValue(Guid.NewGuid().ToString()) },
                                   { "some-random-id-2", new JValue(Guid.NewGuid().ToString()) },
                                   { "some-random-id-3", new JValue(Guid.NewGuid().ToString()) },
                                   { "some-random-id-4", new JValue(Guid.NewGuid().ToString()) },
                                   { "some-random-id-5", new JValue($"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}") },
                                   { "some-random-id-6", new JValue($"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}") },
                                   { "some-random-id-7", new JValue($"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}") },
                                   { "some-random-id-8", new JValue($"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}") },
                                   { "some-random-id-9", new JValue($"{Guid.NewGuid().ToString()}-{Guid.NewGuid().ToString()}") },
                               }
                           },
                           TableUpdateMode.Replace);
```

It fairly simple repro step
1) Start Fiddler with AutoResponder and Rule for Table Storage "PUT" request with delay:2450
![image](https://github.com/Azure/azure-sdk-for-net/assets/5162808/1e47386b-2350-4bb2-9e0e-4fe3df60c823)

2) Run the Console App to start sending the Upsert request to Table Storage.

We also have no idea why the Table SDK end up in this state as we are seeing this in Our Production log intermittently with high probability during higher load. This issue only occurred after we update to "Azure.Data.Tables" from "Microsoft.Azure.Cosmos.Table".

**Comment by christothes** [comment]
Hi @rafflesgun 
I've tried a few ways to reproduce this without success. Since you have a reliable repro with project references to Azure.Core, could you try setting a watch breakpoint on the Request.Content.BytesPending to see what is stepping on the request content? 

![image](https://github.com/Azure/azure-sdk-for-net/assets/1279263/f20038bc-c388-4c9b-94e3-fe4d993c2463)

**Comment by rafflesgun** [comment]
Hi @christothes 
I did that initially (in this thread) to find which `HttpPipelinePolicy` getting empty content after the retry.
And I found that when `OperationCanceledException` throw in `ResponseBodyPolicy`, the Content become empty despite there's no dispose call for `Utf8JsonRequestContent`. 
This will bubble up to `RetryPolicy` for next attempt and from this point onward, the Content will be empty and we start seeing `400 BadRequest - Invalid Input` response as the request is sent with empty Http Body.
![image](https://github.com/Azure/azure-sdk-for-net/assets/5162808/258ff6c6-f618-4a28-9d7f-a16710002032)

**Comment by christothes** [comment]
Hi @rafflesgun - Thanks. Yes, this shows that the request content was emptied sometime between those two steps, but not what is actually emptying it. What I was proposing is to see if you can get the debugger to break on when the request content actually changes.

**Comment by rafflesgun** [comment]
Hi @christothes , I'm getting 0 Bytes for Committed/Pending for `Utf8JsonRequestContent` when it happened. Which doesn't tell what emptying it.
![image](https://github.com/Azure/azure-sdk-for-net/assets/5162808/7954abef-6462-429c-aa6e-03f6efcdf584)

**Comment by christothes** [comment]
Hi @rafflesgun - Did you set a watch breakpoint on the Content value to get it to break on this stack? I suspect it is happening after it leaves the ResponseBodyPolicy.

**Comment by rafflesgun** [comment]
Hi @christothes 
The breakpoint was set in `HttpClientTransport` with additional `catch` statement and the request content already empty when `OperationCanceledException` exception thrown before bubble up.
Before the exception, the request content is **NOT empty** and other than the Request Content changed, I did not notice any other changes.
![image](https://github.com/Azure/azure-sdk-for-net/assets/5162808/3ad6cc70-dd03-44f9-8831-ee151cf4a13d)

**Comment by christothes** [comment]
Hi @rafflesgun 
This seems to indicate that the problem is occurring in the transport (`HttpClient`). Just to confirm, are you setting a **watch breakpoint** specifically for when the Content changes, or just a normal breakpoint in your catch block?

**Comment by rafflesgun** [comment]
Hi @christothes 
I'm using Conditional Breakpoint with `ContentLength == 0` in the catch block.

**Comment by rafflesgun** [comment]
Hi @christothes 
The "Break when value changes" is not working that straightforward as it only applicable for specific variable memory address, unless it set the break on all threads but there's max number of these breakpoint it allows (max 4).
![image](https://github.com/Azure/azure-sdk-for-net/assets/5162808/984faf35-afb1-4c3c-a11a-5e209da59230)

Especially this replication happens intermittent between multiple executions/threads. 
Hence, I'm not able to set breakpoint on variable memory for this issue.

**Comment by christothes** [comment]
I was able to reproduce this finally, by setting the processor affinity on the repro exe to use just 1 processor. Here is what I found:

It appears that when fiddler introduces a delay in the request, it holds the request for the requested delay before forwarding it to the host. In this repro, we are intentionally starving the thread pool which introduces significant delay in the task processing of the requests by the transport, which causes the timeout exceptions. However, in some cases, the timing lines up as so:

- Request sent to transport
- Fiddler starts delay of request send to host
- Request reaches timeout
- Request is disposed
- Fiddler delay period elapses and request is sent to the host

In this sequence, by the time Fiddler sends the actual request, the request content is null because the request has already been disposed.

In summary, this seems like a timing issue related to extreme resource exhaustion.

**Comment by rafflesgun** [comment]
Glad that you able to replicate the issue.
Any possible/potential fix to be done in SDK level? 
As this issue start occurring in production (Azure ASP P2V2 with AutoScale to max 30 instances) after we updated from `Microsoft.Azure.Cosmos.Table v1.0.8` to `Azure.Data.Tables v12.8.0`.

**Comment by christothes** [comment]
What I would recommend is re-thinking the code to avoid as much concurrent threading that causes timeouts, or to relax the timeouts in the options. One way to maximize throughput is to use batching, where possible - see the following sample: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/tables/Azure.Data.Tables/samples/Sample6TransactionalBatch.md
